packages/@lwc/engine/src/framework/invoker.ts

```diff
@@ -104,39 +103,36 @@ export function invokeComponentRenderMethod(vm: VM): VNodes {
         owner,
     } = vm;
     const ctx = currentContext;
-    const isRenderingInception = isRendering;
-    const vmBeingRenderedInception = vmBeingRendered;
-    isRendering = true;
-    vmBeingRendered = vm;
-    let result;
+    const isRenderBeingInvokedInception = isInvokingRender;
+    const vmBeingRenderedInception = getVMBeingRendered();
+    let html: Template | undefined;
     runWithBoundaryProtection(
         vm,
         owner,
         () => {
             // pre
             establishContext(context);
-            if (process.env.NODE_ENV !== 'production') {
-                startMeasure('render', vm);
-            }
-            isRendering = true;
-            vmBeingRendered = vm;
+            isInvokingRender = true;
+            setVMBeingRendered(vm);
         },
         () => {
             // job
-            const html = callHook(component, render);
-            result = evaluateTemplate(vm, html);
+            html = callHook(component, render);
+            if (process.env.NODE_ENV !== 'production') {
+                assert.isTrue(
+                    !isUndefined(html),
+                    `${vm}.render() should return an imported template instead of ${toString(html)}`
+                );
+            }
         },
         () => {
-            establishContext(ctx);
             // post
-            if (process.env.NODE_ENV !== 'production') {
-                endMeasure('render', vm);
-            }
-            isRendering = isRenderingInception;
-            vmBeingRendered = vmBeingRenderedInception;
+            establishContext(ctx);
+            isInvokingRender = isRenderBeingInvokedInception;
+            setVMBeingRendered(vmBeingRenderedInception);
         }
     );
-    return result || [];
+    return isUndefined(html) ? [] : evaluateTemplate(vm, html);
```

Comment:
we could probably make this more clear, maybe adding a comment saying that `html` can only be undefined if there is an error on render.call() that is captured by a boundary.

Reply:
There is a slight inconsistency here. 
`if the render() returns an undefined`
1. In dev mode, we assert and throw, invoke errorCallback in boundary protection
2. In prod mode, we do not throw. We return a blank template to gracefully degrade

`if the render() returns a null/string/non-function`
1. In dev mode, we assert and throw, invoke errorCallback in boundary protection
2. In prod mode, we do not assert, the framework will encounter a TypeError while trying to use the return value like a function. We invoke boundary protection.

I feel, throwing in dev mode and returning a blank template in prod mode is better user experience. Want to hear opinions.

Reply:
this is not the same we have in master today... in master today, if you return undefined, in dev or prod, it throws.

Reply:
in fact it is not about undefined, it is about a registered template.